File: src/types/blog.ts
export interface Blog {
  id: string
  title: string
  content: string
  excerpt: string
  tags: string[]
  authorId: string
  authorName: string
  authorEmail: string
  status: 'draft' | 'published' | 'archived'
  publishedAt?: string
  createdAt: string
  updatedAt: string
  slug: string
  metaTitle?: string
  metaDescription?: string
  featuredImage?: string
  viewCount: number
  likeCount: number
  shareCount: number
}

export interface Comment {
  id: string
  content: string
  authorName: string
  authorEmail: string
  blogId: string
  status: 'pending' | 'approved' | 'rejected'
  createdAt: string
  parentId?: string
  replies?: Comment[]
  blog?: Blog // Include blog information for admin purposes
}

export interface CreateBlogRequest {
  title: string
  content: string
  description: string
  tags: string[]
  status: 'draft' | 'published'
  metaTitle?: string
  metaDescription?: string
  featuredImage?: string
}

export interface AIGenerateRequest {
  title: string
  description: string
  tone: 'professional' | 'casual' | 'technical' | 'friendly'
  length: 'short' | 'medium' | 'long'
}

export interface AIGenerateResponse {
  content: string
  metaTitle: string
  metaDescription: string
  suggestedTags: string[]
} 

File: backend/internal/models/blog.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/lib/pq"
	"gorm.io/gorm"
)

type Blog struct {
	ID              string         `json:"id" gorm:"primaryKey;type:uuid;default:uuid_generate_v4()"`
	Title           string         `json:"title" gorm:"not null"`
	Content         string         `json:"content" gorm:"type:text"`
	Excerpt         string         `json:"excerpt"`
	Slug            string         `json:"slug" gorm:"uniqueIndex;not null"`
	AuthorID        string         `json:"authorId" gorm:"not null"`
	AuthorName      string         `json:"authorName"`
	AuthorEmail     string         `json:"authorEmail"`
	Status          string         `json:"status" gorm:"default:'draft'"` // draft, published, archived
	Tags            pq.StringArray `json:"tags" gorm:"type:text[]"`
	MetaTitle       string         `json:"metaTitle"`
	MetaDescription string         `json:"metaDescription"`
	FeaturedImage   string         `json:"featuredImage"`
	ViewCount       int            `json:"viewCount" gorm:"default:0"`
	LikeCount       int            `json:"likeCount" gorm:"default:0"`
	ShareCount      int            `json:"shareCount" gorm:"default:0"`
	PublishedAt     *time.Time     `json:"publishedAt"`
	CreatedAt       time.Time      `json:"createdAt"`
	UpdatedAt       time.Time      `json:"updatedAt"`
	DeletedAt       gorm.DeletedAt `json:"-" gorm:"index"`
}

func (b *Blog) BeforeCreate(tx *gorm.DB) error {
	if b.ID == "" {
		b.ID = uuid.New().String()
	}
	return nil
}

type CreateBlogRequest struct {
	Title           string   `json:"title" binding:"required"`
	Content         string   `json:"content" binding:"required"`
	Description     string   `json:"description"`
	Tags            []string `json:"tags"`
	Status          string   `json:"status" binding:"required,oneof=draft published"`
	MetaTitle       string   `json:"metaTitle"`
	MetaDescription string   `json:"metaDescription"`
	FeaturedImage   string   `json:"featuredImage"`
}
